### PR TITLE
Remove use of internal gradle APIs

### LINKED ISSUE #149: Investigate removal of usage of internal Gradle APIs
The only internal API that comes to mind yet is `org.gradle.api.internal.file.UnionFileCollection`, which is used in [FormatExtension.java](https://github.com/diffplug/spotless/blob/master/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/FormatExtension.java).

The advantage of moving away from internal APIs, if we can do so, is the Gradle plugin would be more future-proof in the face of internal Gradle changes.

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Closes #149

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
Peachy, I also have to find a replacement for `DslObject`

**Comment by jbduncan** [comment]
:+1:

**Comment by JLLeitschuh** [comment]
You know, this might not be possible.
This is how the Gradle Kotlin DSL solves this problem:
https://github.com/gradle/kotlin-dsl/blob/6160cb88a99f846ad34d916f3847563532c46ac8/buildSrc/src/main/kotlin/build/KotlinPluginExtensions.kt

Notice that they are using `org.gradle.api.internal.HasConvention` to do this.

**Comment by JLLeitschuh** [comment]
Okay, I'll get back to this on Monday. I opened an issue with Gradle about this:
https://github.com/gradle/gradle/issues/3191

File: plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java [hunk 1]
```diff
@@ -128,13 +128,13 @@ protected void setupTask(SpotlessTask task) {
...
+			FileCollection union = getProject().files();
 			for (SourceSet sourceSet : convention.getSourceSets()) {
 				GroovySourceSet groovySourceSet = new DslObject(sourceSet).getConvention().getPlugin(GroovySourceSet.class);
 				if (excludeJava) {
-					union.add(groovySourceSet.getAllGroovy());
+					union = union.plus(groovySourceSet.getAllGroovy());
 				} else {
-					union.add(groovySourceSet.getGroovy());
+					union = union.plus(groovySourceSet.getGroovy());
 				}
```
Comment (nedtwigg) [inline plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java]: Just curious, is there a performance difference here?  Is `FileCollection.plus()` just as "lazy" as `UnionFileCollection`?

Comment (JLLeitschuh) [reply plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java]: The `FileCollection` returned by `getProject().files()` is immutable (as far as I can tell). The only way that a `UnionFileCollection` can be created using the public API is using this `plus` method.

So what's actually happening is that a bunch of `UnionFileCollection`'s are being unioned together.
Don't know if this is really a performance thing besides a bunch of short lived objects being created.

Comment (nedtwigg) [reply plugin-gradle/src/main/java/com/diffplug/gradle/spotless/GroovyExtension.java]: Thanks!  Sounds good enough to me.

**Comment by JLLeitschuh** [comment]
@nedtwigg I don't think there's going to be a good replacement for this line here. Especially in Gradle 2.4
```java
import org.gradle.api.internal.plugins.DslObject;
// In method:
GroovySourceSet groovySourceSet = new DslObject(sourceSet).getConvention().getPlugin(GroovySourceSet.class);
```

I have a few options here:
 - Add an exception to the "spotless self" rules for `import org.gradle.api.internal.plugins.DslObject`
 - Make heavy use of reflection to replace `new DslObject(sourceSet).getConvention().getPlugin(GroovySourceSet.class)` with the reflection based equivalent.

Maybe you have a better solution??

**Comment by nedtwigg** [comment]
The goal in reducing usage of internal API's is reduced chance of breakage.  Whether we access the API via normal function calls or reflection, we'll break just the same, so I don't think reflection helps.

Adding an exception to `spotlessSelf.gradle` seem like a good compromise.  By removing `UnionFileCollection` you've made sure that we won't get burned by changes to that, and `spotlessSelf` makes sure that we don't add anything new accidentally.

**Comment by JLLeitschuh** [comment]
No idea why those tests failed on travis CI. I'm testing locally now.

**Comment by JLLeitschuh** [comment]
Local tests don't fail.
@nedtwigg Can you do me a favor and kick travis to re-run?

**Comment by nedtwigg** [comment]
It re-ran and passed.  LGTM.

**Comment by JLLeitschuh** [comment]
Please feel free to merge at your own discretion.

I don't think the changelog really needs to be updated.

**Comment by nedtwigg** [comment]
👍  Merged

**Comment by JLLeitschuh** [comment]
Hey! Thanks! And thanks for adding me!

**Comment by jbduncan** [comment]
@JLLeitschuh @nedtwigg Thank you very much for taking care of #149 and this PR for me! Although I feel guilty for not addressing #149 myself, I'm very happy that it's finally dealt with. So thanks again. :)

**Pull request merged**